### PR TITLE
perf!: replace error strings by custom errors + use `enum` for `OPERATION_TYPE`

### PR DESCRIPTION
**Pull request by CJ42**
# What does this PR introduce?

## ♻️ Refactor & Perfs 🚀 (⚠️ BREAKING CHANGES)

- ⚠️ replace revert reason strings by custom errors in `ERC725XCore` and `ERC725YCore`. All errors are listed in `errors.sol`.

This reduces the deployment cost of the contracts by around 200,00 gas. See table summary below.

- ⚠️  rename the `uint256` constants for each OPERATION_TYPE by adding the operation number inside the constant name. `OPERATION_CALL` -> `OPERATION_0_CALL`.

- remove modifier to inline owner is not zero address check inside constructor / initializer.

## 🎨 Style

- delete comments that aim to create sections _e.g: public functions, internal functions, overriden functions, etc...)_. Remove these comments in favour of following the Solidity style guide: 

## 📝 Docs

- added Natspec comments for newly created custom errors.
- improve descriptions of existing Natspec docs comments + descriptions of `execute(...)` method in ERC725.md
    - replace `operation` -> `operationType`.
    - specify "low-level" for call operations + "opcode" for contract creation operations.
operationType
    - add "Requirements" section with bullet points under the `execute(...)` function in the `IERC725X` interface contract.
    - change variable name from `data` to `creationCode` for contract deployment internal methods.
    
## 🧪 Tests

- edit all revert reason strings with custom errors.

## 🏗️ Build

- add configs for Hardhat gas reporter.

# Gas Cost improvements

## Deployment Cost

| Contract | Deployment cost before  | Deployment cost after  |  Change (in gas) | Change (in %) |
|---------------|---|---|---|---|
| `ERC725Init`  | 2_554_139  | 2_285_509  | 📉 -268_630 | 📉 -10.52% |
|  `ERC725` |   |  2_075_156 |   |   |
| `ERC725X`    | 1_714_827  | 1_507_680  | 📉 -207_147 | 📉 -12.08% |
| `ERC725XInit`  | 1_956_761  | 1_718_059  | 📉 -238_702 | 📉 -12.20% |
| `ERC725Y`  |  1_084_771 | 1_068_904 | 📉 -15_867 |  📉 -1.46% |
| `ERC725YInit`  | 1_323_811  | 1_276_413  | 📉 -47_398 |  📉 -3.60% |

## Execution Cost

|  Contract  | Method | Gas cost (average) before | Gas cost (average) after | change (in gas) | change (in %) |
|-----------|---------|---------------------------|--------------------------|----------------|---------------|
| `ERC725` | `execute(uint256,address,uint256,bytes)` | 50_903 |  50_720 | 📉 -183 | 📉 -0.36% |
| `ERC725` | `setData(bytes32,bytes)` | 1_468_353  | 1_330_410  | 📉 -137_943 | 📉 -9.39% |
| `ERC725` | `setData (bytes32[],bytes[])` |  1_393_483 | 1_263_203  | 📉 -130_280 | 📉 -9.35% |
| `ERC725X` | `execute(uint256,address,uint256,bytes)` | 48_141  | 47_958  | 📉 -183  | 📉 -0.38%%  |
| `ERC725Y` | `setData (bytes32, bytes)` | 1_465_204  |  1_327_305 | 📉 -137_899  | 📉 -9.41% |
| `ERC725Y` | `setData (bytes32[],bytes[])` | 1_390_330 | 1_260_092 | 📉 -130_238 | 📉 -9.37%  |

### CODE REVIEW

**Review by YamenMerhi — CHANGES_REQUESTED** [review]
Looks good to me, but I would prefer to revert the last commit to keep the checks inside the internal functions. 

The main goal of the internal function is to have every logic + every check of the operation to make inside the relevant internal function, for context reasons, and for future inheritance reasons.

Minor pt:
I prefer the one-line syntax for the revert (if it fits on one line). example:
`if (data.length == 0) revert ERC725X_NoContractBytecodeProvided();` 

instead of:
```
if (data.length == 0) {
     revert ERC725X_NoContractBytecodeProvided();  
}
```

**Comment by CJ42** [comment]
> Looks good to me, but I would prefer to revert the last commit to keep the checks inside the internal functions.

The reason I refactored here in this last commit is two fold, First because it found out that it saves 1,309 gas of deployment cost (The contract ends up with more opcodes, likely because there are more stack operations to pass more arguments down to the internal functions). My second reason is explained below.

> The main goal of the internal function is to have every logic + every check of the operation to make inside the relevant internal function, for context reasons, and for future inheritance reasons.

Could you give a bit more the "context and inheritance reasons"?

These my arguments for and against the last commit.

**For reverting the last commit**
- I do agree that the one line `if` statement make it a bit nicer to read. 

**Against reverting the last commit**
- if you look at the refactoring of the internal functions in the last commit, the 3 internal functions for each message calls (`_executeCall(...)`, `_executeStaticCall(...)` and `_executeDelegateCall(...)` now look all the same inside (screenshot below).
- it sounds quite contradictory to create an internal function with a parameter, but this parameter can never be used, because it should always be the same ( `value = 0` for msg calls, and `to = address(0)` for contract deployment). For a dev using the internal function only (_e.g: `_executeStaticCall(...)`), why should it pass an argument to the function if this argument is not used? (here `value` for staticcall).
- since these functions are `virtual`, can't someone have the risk forget to add this check if it overrides?

image

**Comment by CJ42** [comment]
@YamenMerhi 

If you look at the equivalent functions from the OpenZeppelin `Address` library , the internal functions `functionStaticCall(...)` and `functionDelegateCall(...)` do not have the `value` parameter.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/a1948250ab8c441f6d327a65754cb20d2b1b4554/contracts/utils/Address.sol#L139-L187

File: implementations/contracts/ERC725XCore.sol [hunk 1]
```diff
@@ -110,108 +120,94 @@ abstract contract ERC725XCore is OwnableUnset, ERC165, IERC725X {
         uint256 value,
         bytes memory data
     ) internal virtual returns (bytes memory result) {
-        emit Executed(OPERATION_CALL, to, value, bytes4(data));
+        emit Executed(uint256(OPERATION_TYPE.CALL), to, value, bytes4(data));
 
         // solhint-disable avoid-low-level-calls
         (bool success, bytes memory returnData) = to.call{value: value}(data);
         result = Address.verifyCallResult(success, returnData, "ERC725X: Unknown Error");
```
Comment (CJ42) [inline implementations/contracts/ERC725XCore.sol:127]: @YamenMerhi I am happy to revert back the last commit if we decide to keep the old structure of the internal functions.

The only extra change I would suggest is to change the revert error message passed to the OZ `Address.sol` library to use the same one they have.

So replace for instance `ERC725X: Unknown Error` > `ERC725X: low-level static call failed`

Screenshot 2022-10-21 at 00 23 04

**Comment by YamenMerhi** [comment]
My main idea was to have every check related to the execution in the internal function so that later if someone wants to implement these execution features they have the checks already in and don't need to add them again. For instance in LSP9Vault, then we need to re-write the logic in the main function again like this: 

```js
if (operationType == uint256(OPERATION_TYPE.STATICCALL)) {
            if (value != 0) revert ERC725X_MsgValueDisallowedInStaticCall();
            return _executeStaticCall(to, data);
        }
        
        if (operationType == uint256(OPERATION_TYPE.CALL)) {
            return _executeCall(to, value, data);
        }
        

        if (operationType == uint256(OPERATION_TYPE.CREATE)) {
            if (to != address(0)) revert ERC725X_CreateOperationsRequireEmptyRecipientAddress();
            return _deployCreate(value, data);
        }

```

instead of doing this:

```js
     If (operationType == uint256(OPERATION_TYPE.STATICCALL)) return _executeStaticCall(to, data);
        
        if (operationType == uint256(OPERATION_TYPE.CALL)) return _executeCall(to, value, data);
        

        if (operationType == uint256(OPERATION_TYPE.CREATE)) return _deployCreate(value, data);

```
But in the end, it's a minor pt, I am happy also to keep it like you change it now.

**Review by YamenMerhi — APPROVED** [review]

**Review by skimaharvey — APPROVED** [review]

**Review by frozeman — APPROVED** [review]

**Comment by CJ42** [comment]
As discussed, we remove the `enum` and change back to `OPERATION_0_CALL`, ... (the operation number in the constant name).

**Pull request merged**